// @vue-mini/core@1.2.0/dist/vue-mini.esm-bundler.js
/*!
 * vue-mini v1.2.0
 * https://github.com/vue-mini/vue-mini
 * (c) 2019-present Yang Mingshan
 * @license MIT
 */
import { watch as watch$1, isRef, isProxy, toRaw, EffectScope, shallowReactive, shallowReadonly } from '@vue/reactivity';
export { EffectScope, ReactiveEffect, TrackOpTypes, TriggerOpTypes, computed, customRef, effect, effectScope, getCurrentScope, getCurrentWatcher, isProxy, isReactive, isReadonly, isRef, isShallow, markRaw, onScopeDispose, onWatcherCleanup, proxyRefs, reactive, readonly, ref, shallowReactive, shallowReadonly, shallowRef, stop, toRaw, toRef, toRefs, toValue, triggerRef, unref } from '@vue/reactivity';

const EMPTY_OBJ = (process.env.NODE_ENV !== 'production') ? Object.freeze({}) : /* istanbul ignore next -- @preserve */ {};
/* istanbul ignore next -- @preserve */
// eslint-disable-next-line @typescript-eslint/no-empty-function
const NOOP = () => { };
const { isArray } = Array;
const extend = Object.assign;
function exclude(obj, keys) {
    const ret = {};
    Object.keys(obj).forEach((key) => {
        if (!keys.includes(key)) {
            ret[key] = obj[key];
        }
    });
    return ret;
}
function getType(x) {
    return Object.prototype.toString.call(x).slice(8, -1);
}
function isSimpleValue(x) {
    const simpleTypes = new Set(['undefined', 'boolean', 'number', 'string']);
    return x === null || simpleTypes.has(typeof x);
}
function isObject(x) {
    return x !== null && typeof x === 'object';
}
function isPlainObject(x) {
    return getType(x) === 'Object';
}
function isFunction(x) {
    return typeof x === 'function';
}
function toHiddenField(name) {
    return `__${name}__`;
}

/* eslint-disable no-bitwise, unicorn/prefer-math-trunc, @typescript-eslint/prefer-literal-enum-member */
var SchedulerJobFlags;
(function (SchedulerJobFlags) {
    SchedulerJobFlags[SchedulerJobFlags["QUEUED"] = 1] = "QUEUED";
    SchedulerJobFlags[SchedulerJobFlags["ALLOW_RECURSE"] = 4] = "ALLOW_RECURSE";
})(SchedulerJobFlags || (SchedulerJobFlags = {}));
const queue = [];
let flushIndex = -1;
const pendingPostFlushCbs = [];
let activePostFlushCbs = null;
let postFlushIndex = 0;
// eslint-disable-next-line spaced-comment
const resolvedPromise = /*@__PURE__*/ Promise.resolve();
let currentFlushPromise = null;
const RECURSION_LIMIT = 100;
function nextTick(fn) {
    const p = currentFlushPromise || resolvedPromise;
    // eslint-disable-next-line promise/prefer-await-to-then
    return fn ? p.then(fn) : p;
}
function queueJob(job) {
    if (!(job.flags & SchedulerJobFlags.QUEUED)) {
        queue.push(job);
        job.flags |= SchedulerJobFlags.QUEUED;
        queueFlush();
    }
}
function queueFlush() {
    if (!currentFlushPromise) {
        // eslint-disable-next-line promise/prefer-await-to-then
        currentFlushPromise = resolvedPromise.then(flushJobs);
    }
}
function queuePostFlushCb(cb) {
    if (!(cb.flags & SchedulerJobFlags.QUEUED)) {
        pendingPostFlushCbs.push(cb);
        cb.flags |= SchedulerJobFlags.QUEUED;
    }
}
function flushPostFlushCbs() {
    if (pendingPostFlushCbs.length > 0) {
        activePostFlushCbs = [...new Set(pendingPostFlushCbs)];
        pendingPostFlushCbs.length = 0;
        for (postFlushIndex = 0; postFlushIndex < activePostFlushCbs.length; postFlushIndex++) {
            const cb = activePostFlushCbs[postFlushIndex];
            if (cb.flags & SchedulerJobFlags.ALLOW_RECURSE) {
                cb.flags &= ~SchedulerJobFlags.QUEUED;
            }
            cb();
            cb.flags &= ~SchedulerJobFlags.QUEUED;
        }
        activePostFlushCbs = null;
        postFlushIndex = 0;
    }
}
function flushJobs(seen) {
    /* istanbul ignore else -- @preserve  */
    if ((process.env.NODE_ENV !== 'production')) {
        seen = seen || new Map();
    }
    // Conditional usage of checkRecursiveUpdate must be determined out of
    // try ... catch block since Rollup by default de-optimizes treeshaking
    // inside try-catch. This can leave all warning code unshaked. Although
    // they would get eventually shaken by a minifier like terser, some minifiers
    // would fail to do that (e.g. https://github.com/evanw/esbuild/issues/1610)
    const check = (process.env.NODE_ENV !== 'production') ?
        (job) => checkRecursiveUpdates(seen, job)
        : /* istanbul ignore next -- @preserve  */ NOOP;
    try {
        for (flushIndex = 0; flushIndex < queue.length; flushIndex++) {
            const job = queue[flushIndex];
            /* istanbul ignore if -- @preserve  */
            if ((process.env.NODE_ENV !== 'production') && check(job)) {
                continue;
            }
            if (job.flags & SchedulerJobFlags.ALLOW_RECURSE) {
                job.flags &= ~SchedulerJobFlags.QUEUED;
            }
            job();
            if (!(job.flags & SchedulerJobFlags.ALLOW_RECURSE)) {
                job.flags &= ~SchedulerJobFlags.QUEUED;
            }
        }
    }
    finally {
        // If there was an error we still need to clear the QUEUED flags
        for (; flushIndex < queue.length; flushIndex++) {
            const job = queue[flushIndex];
            job.flags &= ~SchedulerJobFlags.QUEUED;
        }
        flushIndex = -1;
        queue.length = 0;
        currentFlushPromise = null;
    }
}
function checkRecursiveUpdates(seen, fn) {
    const count = seen.get(fn) || 0;
    /* istanbul ignore if -- @preserve */
    if (count > RECURSION_LIMIT) {
        console.warn(`Maximum recursive updates exceeded. ` +
            `This means you have a reactive effect that is mutating its own ` +
            `dependencies and thus recursively triggering itself.`);
        return true;
    }
    seen.set(fn, count + 1);
    return false;
}

// Simple effect.
function watchEffect(effect, options) {
    return doWatch(effect, null, options);
}
function watchPostEffect(effect, options) {
    return doWatch(effect, null, (process.env.NODE_ENV !== 'production') ?
        extend({}, options, { flush: 'post' })
        : /* istanbul ignore next -- @preserve */ { flush: 'post' });
}
function watchSyncEffect(effect, options) {
    return doWatch(effect, null, (process.env.NODE_ENV !== 'production') ?
        extend({}, options, { flush: 'sync' })
        : /* istanbul ignore next -- @preserve */ { flush: 'sync' });
}
// Implementation
function watch(source, cb, options) {
    if ((process.env.NODE_ENV !== 'production') && !isFunction(cb)) {
        console.warn(`\`watch(fn, options?)\` signature has been moved to a separate API. ` +
            `Use \`watchEffect(fn, options?)\` instead. \`watch\` now only ` +
            `supports \`watch(source, cb, options?) signature.`);
    }
    return doWatch(source, cb, options);
}
function doWatch(source, cb, options = EMPTY_OBJ) {
    const { immediate, deep, flush, once } = options;
    if ((process.env.NODE_ENV !== 'production') && !cb) {
        if (immediate !== undefined) {
            console.warn(`watch() "immediate" option is only respected when using the ` +
                `watch(source, callback, options?) signature.`);
        }
        if (deep !== undefined) {
            console.warn(`watch() "deep" option is only respected when using the ` +
                `watch(source, callback, options?) signature.`);
        }
        if (once !== undefined) {
            console.warn(`watch() "once" option is only respected when using the ` +
                `watch(source, callback, options?) signature.`);
        }
    }
    const baseWatchOptions = extend({}, options);
    // Scheduler
    if (flush === 'post') {
        baseWatchOptions.scheduler = (job) => {
            queuePostFlushCb(job);
        };
    }
    else if (flush !== 'sync') {
        baseWatchOptions.scheduler = (job, isFirstRun) => {
            if (isFirstRun) {
                job();
            }
            else {
                queueJob(job);
            }
        };
    }
    // @ts-expect-error
    baseWatchOptions.augmentJob = (job) => {
        // Important: mark the job as a watcher callback so that scheduler knows
        // it is allowed to self-trigger
        if (cb) {
            // eslint-disable-next-line no-bitwise
            job.flags |= SchedulerJobFlags.ALLOW_RECURSE;
        }
    };
    const watchHandle = watch$1(source, cb, baseWatchOptions);
    return watchHandle;
}

const provides = Object.create(null);
function provide(key, value) {
    // TS doesn't allow symbol as index type
    provides[key] = value;
}
function inject(key, defaultValue, treatDefaultAsFactory = false) {
    if (key in provides) {
        // TS doesn't allow symbol as index type
        return provides[key];
    }
    if (arguments.length > 1) {
        return treatDefaultAsFactory && isFunction(defaultValue) ? defaultValue()
            : defaultValue;
    }
    /* istanbul ignore else -- @preserve */
    if ((process.env.NODE_ENV !== 'production')) {
        console.warn(`injection "${String(key)}" not found.`);
    }
}

let currentApp = null;
let currentPage = null;
let currentComponent = null;
function getCurrentInstance() {
    return currentPage || currentComponent;
}
function setCurrentApp(app) {
    currentApp = app;
}
function unsetCurrentApp() {
    currentApp = null;
}
function setCurrentPage(page) {
    currentPage = page;
    // @ts-expect-error
    page.__scope__.on();
}
function unsetCurrentPage() {
    /* istanbul ignore else -- @preserve */
    if (currentPage) {
        // @ts-expect-error
        currentPage.__scope__.off();
    }
    currentPage = null;
}
function setCurrentComponent(component) {
    currentComponent = component;
    // @ts-expect-error
    component.__scope__.on();
}
function unsetCurrentComponent() {
    /* istanbul ignore else -- @preserve */
    if (currentComponent) {
        // @ts-expect-error
        currentComponent.__scope__.off();
    }
    currentComponent = null;
}

var AppLifecycle;
(function (AppLifecycle) {
    AppLifecycle["ON_LAUNCH"] = "onLaunch";
    AppLifecycle["ON_SHOW"] = "onShow";
    AppLifecycle["ON_HIDE"] = "onHide";
    AppLifecycle["ON_ERROR"] = "onError";
    AppLifecycle["ON_PAGE_NOT_FOUND"] = "onPageNotFound";
    AppLifecycle["ON_UNHANDLED_REJECTION"] = "onUnhandledRejection";
    AppLifecycle["ON_THEME_CHANGE"] = "onThemeChange";
})(AppLifecycle || (AppLifecycle = {}));
function createApp(optionsOrSetup) {
    let setup;
    let options;
    if (isFunction(optionsOrSetup)) {
        setup = optionsOrSetup;
        options = {};
    }
    else {
        if (optionsOrSetup.setup === undefined) {
            // eslint-disable-next-line new-cap
            App(optionsOrSetup);
            return;
        }
        setup = optionsOrSetup.setup;
        options = exclude(optionsOrSetup, ['setup']);
    }
    const originOnLaunch = options[AppLifecycle.ON_LAUNCH];
    options[AppLifecycle.ON_LAUNCH] = function (options) {
        setCurrentApp(this);
        const bindings = setup(options);
        if (bindings !== undefined) {
            Object.keys(bindings).forEach((key) => {
                this[key] = bindings[key];
            });
        }
        unsetCurrentApp();
        if (originOnLaunch !== undefined) {
            originOnLaunch.call(this, options);
        }
    };
    options[AppLifecycle.ON_SHOW] = createLifecycle$2(options, AppLifecycle.ON_SHOW);
    options[AppLifecycle.ON_HIDE] = createLifecycle$2(options, AppLifecycle.ON_HIDE);
    options[AppLifecycle.ON_ERROR] = createLifecycle$2(options, AppLifecycle.ON_ERROR);
    options[AppLifecycle.ON_PAGE_NOT_FOUND] = createLifecycle$2(options, AppLifecycle.ON_PAGE_NOT_FOUND);
    options[AppLifecycle.ON_UNHANDLED_REJECTION] = createLifecycle$2(options, AppLifecycle.ON_UNHANDLED_REJECTION);
    options[AppLifecycle.ON_THEME_CHANGE] = createLifecycle$2(options, AppLifecycle.ON_THEME_CHANGE);
    // eslint-disable-next-line new-cap
    App(options);
}
function createLifecycle$2(options, lifecycle) {
    const originLifecycle = options[lifecycle];
    return function (...args) {
        const hooks = this[toHiddenField(lifecycle)];
        if (hooks) {
            hooks.forEach((hook) => hook(...args));
        }
        if (originLifecycle !== undefined) {
            originLifecycle.call(this, ...args);
        }
    };
}

function deepToRaw(x) {
    if (isSimpleValue(x) || isFunction(x)) {
        return x;
    }
    if (isRef(x)) {
        return deepToRaw(x.value);
    }
    if (isProxy(x)) {
        return deepToRaw(toRaw(x));
    }
    if (isArray(x)) {
        return x.map((item) => deepToRaw(item));
    }
    if (isPlainObject(x)) {
        const obj = {};
        Object.keys(x).forEach((key) => {
            obj[key] = deepToRaw(x[key]);
        });
        return obj;
    }
    throw new TypeError(`${getType(x)} value is not supported`);
}
function deepWatch(key, value) {
    if (!isObject(value)) {
        return;
    }
    watch(isRef(value) ? value : () => value, () => {
        this.setData({ [key]: deepToRaw(value) }, flushPostFlushCbs);
    }, {
        deep: true,
    });
}

var PageLifecycle;
(function (PageLifecycle) {
    PageLifecycle["ON_LOAD"] = "onLoad";
    PageLifecycle["ON_SHOW"] = "onShow";
    PageLifecycle["ON_READY"] = "onReady";
    PageLifecycle["ON_HIDE"] = "onHide";
    PageLifecycle["ON_UNLOAD"] = "onUnload";
    PageLifecycle["ON_ROUTE_DONE"] = "onRouteDone";
    PageLifecycle["ON_PULL_DOWN_REFRESH"] = "onPullDownRefresh";
    PageLifecycle["ON_REACH_BOTTOM"] = "onReachBottom";
    PageLifecycle["ON_PAGE_SCROLL"] = "onPageScroll";
    PageLifecycle["ON_SHARE_APP_MESSAGE"] = "onShareAppMessage";
    PageLifecycle["ON_SHARE_TIMELINE"] = "onShareTimeline";
    PageLifecycle["ON_ADD_TO_FAVORITES"] = "onAddToFavorites";
    PageLifecycle["ON_RESIZE"] = "onResize";
    PageLifecycle["ON_TAB_ITEM_TAP"] = "onTabItemTap";
    PageLifecycle["ON_SAVE_EXIT_STATE"] = "onSaveExitState";
})(PageLifecycle || (PageLifecycle = {}));
function definePage(optionsOrSetup, config) {
    config = extend({
        listenPageScroll: false,
        canShareToOthers: false,
        canShareToTimeline: false,
    }, config);
    let setup;
    let options;
    if (isFunction(optionsOrSetup)) {
        setup = optionsOrSetup;
        options = {};
    }
    else {
        if (optionsOrSetup.setup === undefined) {
            // eslint-disable-next-line new-cap
            Page(optionsOrSetup);
            return;
        }
        setup = optionsOrSetup.setup;
        options = exclude(optionsOrSetup, ['setup']);
    }
    const originOnLoad = options[PageLifecycle.ON_LOAD];
    options[PageLifecycle.ON_LOAD] = function (query) {
        this.__scope__ = new EffectScope();
        setCurrentPage(this);
        const context = {
            is: this.is,
            route: this.route,
            options: this.options,
            exitState: this.exitState,
            router: this.router,
            pageRouter: this.pageRouter,
            renderer: this.renderer,
            createSelectorQuery: this.createSelectorQuery.bind(this),
            createIntersectionObserver: this.createIntersectionObserver.bind(this),
            createMediaQueryObserver: this.createMediaQueryObserver.bind(this),
            selectComponent: this.selectComponent.bind(this),
            selectAllComponents: this.selectAllComponents.bind(this),
            getTabBar: this.getTabBar.bind(this),
            getPageId: this.getPageId.bind(this),
            animate: this.animate.bind(this),
            clearAnimation: this.clearAnimation.bind(this),
            getOpenerEventChannel: this.getOpenerEventChannel.bind(this),
            applyAnimatedStyle: this.applyAnimatedStyle.bind(this),
            clearAnimatedStyle: this.clearAnimatedStyle.bind(this),
            setUpdatePerformanceListener: this.setUpdatePerformanceListener.bind(this),
            getPassiveEvent: this.getPassiveEvent.bind(this),
            setPassiveEvent: this.setPassiveEvent.bind(this),
        };
        const bindings = setup(query, context);
        if (bindings !== undefined) {
            let data;
            Object.keys(bindings).forEach((key) => {
                const value = bindings[key];
                if (isFunction(value)) {
                    this[key] = value;
                    return;
                }
                data = data || {};
                data[key] = deepToRaw(value);
                deepWatch.call(this, key, value);
            });
            if (data !== undefined) {
                this.setData(data, flushPostFlushCbs);
            }
        }
        unsetCurrentPage();
        if (originOnLoad !== undefined) {
            originOnLoad.call(this, query);
        }
    };
    const onUnload = createLifecycle$1(options, PageLifecycle.ON_UNLOAD);
    options[PageLifecycle.ON_UNLOAD] = function () {
        onUnload.call(this);
        this.__scope__.stop();
    };
    if (options[PageLifecycle.ON_PAGE_SCROLL] || config.listenPageScroll) {
        options[PageLifecycle.ON_PAGE_SCROLL] = createLifecycle$1(options, PageLifecycle.ON_PAGE_SCROLL);
        /* istanbul ignore next -- @preserve */
        options.__listenPageScroll__ = () => true;
    }
    if (options[PageLifecycle.ON_SHARE_APP_MESSAGE] === undefined &&
        config.canShareToOthers) {
        options[PageLifecycle.ON_SHARE_APP_MESSAGE] = function (share) {
            const hook = this[toHiddenField(PageLifecycle.ON_SHARE_APP_MESSAGE)];
            if (hook) {
                return hook(share);
            }
            return {};
        };
        /* istanbul ignore next -- @preserve */
        options.__isInjectedShareToOthersHook__ = () => true;
    }
    if (options[PageLifecycle.ON_SHARE_TIMELINE] === undefined &&
        config.canShareToTimeline) {
        options[PageLifecycle.ON_SHARE_TIMELINE] = function () {
            const hook = this[toHiddenField(PageLifecycle.ON_SHARE_TIMELINE)];
            if (hook) {
                return hook();
            }
            return {};
        };
        /* istanbul ignore next -- @preserve */
        options.__isInjectedShareToTimelineHook__ = () => true;
    }
    if (options[PageLifecycle.ON_ADD_TO_FAVORITES] === undefined) {
        options[PageLifecycle.ON_ADD_TO_FAVORITES] = function (favorites) {
            const hook = this[toHiddenField(PageLifecycle.ON_ADD_TO_FAVORITES)];
            if (hook) {
                return hook(favorites);
            }
            return {};
        };
        /* istanbul ignore next -- @preserve */
        options.__isInjectedFavoritesHook__ = () => true;
    }
    if (options[PageLifecycle.ON_SAVE_EXIT_STATE] === undefined) {
        options[PageLifecycle.ON_SAVE_EXIT_STATE] = function () {
            const hook = this[toHiddenField(PageLifecycle.ON_SAVE_EXIT_STATE)];
            if (hook) {
                return hook();
            }
            return { data: undefined };
        };
        /* istanbul ignore next -- @preserve */
        options.__isInjectedExitStateHook__ = () => true;
    }
    options[PageLifecycle.ON_SHOW] = createLifecycle$1(options, PageLifecycle.ON_SHOW);
    options[PageLifecycle.ON_READY] = createLifecycle$1(options, PageLifecycle.ON_READY);
    options[PageLifecycle.ON_HIDE] = createLifecycle$1(options, PageLifecycle.ON_HIDE);
    options[PageLifecycle.ON_ROUTE_DONE] = createLifecycle$1(options, PageLifecycle.ON_ROUTE_DONE);
    options[PageLifecycle.ON_PULL_DOWN_REFRESH] = createLifecycle$1(options, PageLifecycle.ON_PULL_DOWN_REFRESH);
    options[PageLifecycle.ON_REACH_BOTTOM] = createLifecycle$1(options, PageLifecycle.ON_REACH_BOTTOM);
    options[PageLifecycle.ON_RESIZE] = createLifecycle$1(options, PageLifecycle.ON_RESIZE);
    options[PageLifecycle.ON_TAB_ITEM_TAP] = createLifecycle$1(options, PageLifecycle.ON_TAB_ITEM_TAP);
    // eslint-disable-next-line new-cap
    Page(options);
}
function createLifecycle$1(options, lifecycle) {
    const originLifecycle = options[lifecycle];
    return function (...args) {
        const hooks = this[toHiddenField(lifecycle)];
        if (hooks) {
            hooks.forEach((hook) => hook(...args));
        }
        if (originLifecycle !== undefined) {
            originLifecycle.call(this, ...args);
        }
    };
}

var ComponentLifecycle;
(function (ComponentLifecycle) {
    ComponentLifecycle["ATTACHED"] = "attached";
    ComponentLifecycle["READY"] = "ready";
    ComponentLifecycle["MOVED"] = "moved";
    ComponentLifecycle["DETACHED"] = "detached";
    ComponentLifecycle["ERROR"] = "error";
})(ComponentLifecycle || (ComponentLifecycle = {}));
const SpecialLifecycleMap = {
    [PageLifecycle.ON_SHOW]: 'show',
    [PageLifecycle.ON_HIDE]: 'hide',
    [PageLifecycle.ON_RESIZE]: 'resize',
    [PageLifecycle.ON_ROUTE_DONE]: 'routeDone',
    [ComponentLifecycle.READY]: PageLifecycle.ON_READY,
};
function defineComponent(optionsOrSetup, config) {
    config = extend({
        listenPageScroll: false,
        canShareToOthers: false,
        canShareToTimeline: false,
    }, config);
    let setup;
    let options;
    let properties = null;
    if (isFunction(optionsOrSetup)) {
        setup = optionsOrSetup;
        options = {};
    }
    else {
        if (optionsOrSetup.setup === undefined) {
            // eslint-disable-next-line new-cap
            return Component(optionsOrSetup);
        }
        setup = optionsOrSetup.setup;
        options = exclude(optionsOrSetup, ['setup']);
        if (options.properties) {
            properties = Object.keys(options.properties);
        }
    }
    if (options.lifetimes === undefined) {
        options.lifetimes = {};
    }
    const originAttached = options.lifetimes[ComponentLifecycle.ATTACHED] ||
        options[ComponentLifecycle.ATTACHED];
    options.lifetimes[ComponentLifecycle.ATTACHED] = function () {
        this.__scope__ = new EffectScope();
        setCurrentComponent(this);
        const rawProps = {};
        if (properties) {
            properties.forEach((property) => {
                rawProps[property] = this.data[property];
            });
        }
        this.__props__ = shallowReactive(rawProps);
        const context = {
            is: this.is,
            id: this.id,
            dataset: this.dataset,
            exitState: this.exitState,
            router: this.router,
            pageRouter: this.pageRouter,
            renderer: this.renderer,
            triggerEvent: this.triggerEvent.bind(this),
            createSelectorQuery: this.createSelectorQuery.bind(this),
            createIntersectionObserver: this.createIntersectionObserver.bind(this),
            createMediaQueryObserver: this.createMediaQueryObserver.bind(this),
            selectComponent: this.selectComponent.bind(this),
            selectAllComponents: this.selectAllComponents.bind(this),
            selectOwnerComponent: this.selectOwnerComponent.bind(this),
            getRelationNodes: this.getRelationNodes.bind(this),
            getTabBar: this.getTabBar.bind(this),
            getPageId: this.getPageId.bind(this),
            animate: this.animate.bind(this),
            clearAnimation: this.clearAnimation.bind(this),
            getOpenerEventChannel: this.getOpenerEventChannel.bind(this),
            applyAnimatedStyle: this.applyAnimatedStyle.bind(this),
            clearAnimatedStyle: this.clearAnimatedStyle.bind(this),
            setUpdatePerformanceListener: this.setUpdatePerformanceListener.bind(this),
            getPassiveEvent: this.getPassiveEvent.bind(this),
            setPassiveEvent: this.setPassiveEvent.bind(this),
        };
        const bindings = setup((process.env.NODE_ENV !== 'production') ?
            shallowReadonly(this.__props__)
            : /* istanbul ignore next -- @preserve */ this.__props__, context);
        if (bindings !== undefined) {
            let data;
            Object.keys(bindings).forEach((key) => {
                const value = bindings[key];
                if (isFunction(value)) {
                    this[key] = value;
                    return;
                }
                data = data || {};
                data[key] = deepToRaw(value);
                deepWatch.call(this, key, value);
            });
            if (data !== undefined) {
                this.setData(data, flushPostFlushCbs);
            }
        }
        unsetCurrentComponent();
        if (originAttached !== undefined) {
            originAttached.call(this);
        }
    };
    const detached = createComponentLifecycle(options, ComponentLifecycle.DETACHED);
    options.lifetimes[ComponentLifecycle.DETACHED] = function () {
        detached.call(this);
        this.__scope__.stop();
    };
    const originReady = options.lifetimes[ComponentLifecycle.READY] ||
        options[ComponentLifecycle.READY];
    options.lifetimes[ComponentLifecycle.READY] = createLifecycle(SpecialLifecycleMap[ComponentLifecycle.READY], originReady);
    options.lifetimes[ComponentLifecycle.MOVED] = createComponentLifecycle(options, ComponentLifecycle.MOVED);
    options.lifetimes[ComponentLifecycle.ERROR] = createComponentLifecycle(options, ComponentLifecycle.ERROR);
    if (options.methods === undefined) {
        options.methods = {};
    }
    if (options.methods[PageLifecycle.ON_PAGE_SCROLL] ||
        config.listenPageScroll) {
        options.methods[PageLifecycle.ON_PAGE_SCROLL] = createPageLifecycle(options, PageLifecycle.ON_PAGE_SCROLL);
        /* istanbul ignore next -- @preserve */
        options.methods.__listenPageScroll__ = () => true;
    }
    if (options.methods[PageLifecycle.ON_SHARE_APP_MESSAGE] === undefined &&
        config.canShareToOthers) {
        options.methods[PageLifecycle.ON_SHARE_APP_MESSAGE] = function (share) {
            const hook = this[toHiddenField(PageLifecycle.ON_SHARE_APP_MESSAGE)];
            if (hook) {
                return hook(share);
            }
            return {};
        };
        /* istanbul ignore next -- @preserve */
        options.methods.__isInjectedShareToOthersHook__ = () => true;
    }
    if (options.methods[PageLifecycle.ON_SHARE_TIMELINE] === undefined &&
        config.canShareToTimeline) {
        options.methods[PageLifecycle.ON_SHARE_TIMELINE] = function () {
            const hook = this[toHiddenField(PageLifecycle.ON_SHARE_TIMELINE)];
            if (hook) {
                return hook();
            }
            return {};
        };
        /* istanbul ignore next -- @preserve */
        options.methods.__isInjectedShareToTimelineHook__ = () => true;
    }
    if (options.methods[PageLifecycle.ON_ADD_TO_FAVORITES] === undefined) {
        options.methods[PageLifecycle.ON_ADD_TO_FAVORITES] = function (favorites) {
            const hook = this[toHiddenField(PageLifecycle.ON_ADD_TO_FAVORITES)];
            if (hook) {
                return hook(favorites);
            }
            return {};
        };
        /* istanbul ignore next -- @preserve */
        options.methods.__isInjectedFavoritesHook__ = () => true;
    }
    if (options.methods[PageLifecycle.ON_SAVE_EXIT_STATE] === undefined) {
        options.methods[PageLifecycle.ON_SAVE_EXIT_STATE] = function () {
            const hook = this[toHiddenField(PageLifecycle.ON_SAVE_EXIT_STATE)];
            if (hook) {
                return hook();
            }
            return { data: undefined };
        };
        /* istanbul ignore next -- @preserve */
        options.methods.__isInjectedExitStateHook__ = () => true;
    }
    options.methods[PageLifecycle.ON_LOAD] = createPageLifecycle(options, PageLifecycle.ON_LOAD);
    options.methods[PageLifecycle.ON_PULL_DOWN_REFRESH] = createPageLifecycle(options, PageLifecycle.ON_PULL_DOWN_REFRESH);
    options.methods[PageLifecycle.ON_REACH_BOTTOM] = createPageLifecycle(options, PageLifecycle.ON_REACH_BOTTOM);
    options.methods[PageLifecycle.ON_TAB_ITEM_TAP] = createPageLifecycle(options, PageLifecycle.ON_TAB_ITEM_TAP);
    if (options.pageLifetimes === undefined) {
        options.pageLifetimes = {};
    }
    options.pageLifetimes[SpecialLifecycleMap[PageLifecycle.ON_SHOW]] =
        createSpecialPageLifecycle(options, PageLifecycle.ON_SHOW);
    options.pageLifetimes[SpecialLifecycleMap[PageLifecycle.ON_HIDE]] =
        createSpecialPageLifecycle(options, PageLifecycle.ON_HIDE);
    options.pageLifetimes[SpecialLifecycleMap[PageLifecycle.ON_RESIZE]] =
        createSpecialPageLifecycle(options, PageLifecycle.ON_RESIZE);
    options.pageLifetimes[SpecialLifecycleMap[PageLifecycle.ON_ROUTE_DONE]] =
        createSpecialPageLifecycle(options, PageLifecycle.ON_ROUTE_DONE);
    if (properties) {
        if (options.observers === undefined) {
            options.observers = {};
        }
        properties.forEach((property) => {
            const originObserver = options.observers[property];
            options.observers[property] = function (value) {
                // Observer executes before attached
                if (this.__props__) {
                    this.__props__[property] = value;
                }
                if (originObserver !== undefined) {
                    originObserver.call(this, value);
                }
            };
        });
    }
    // eslint-disable-next-line new-cap
    return Component(options);
}
function createComponentLifecycle(options, lifecycle) {
    const originLifecycle = options.lifetimes[lifecycle] || options[lifecycle];
    return createLifecycle(lifecycle, originLifecycle);
}
function createPageLifecycle(options, lifecycle) {
    const originLifecycle = options.methods[lifecycle];
    return createLifecycle(lifecycle, originLifecycle);
}
function createSpecialPageLifecycle(options, lifecycle) {
    const originLifecycle = options.pageLifetimes[SpecialLifecycleMap[lifecycle]];
    return createLifecycle(lifecycle, originLifecycle);
}
function createLifecycle(lifecycle, originLifecycle) {
    const hiddenField = toHiddenField(lifecycle);
    return function (...args) {
        const hooks = this[hiddenField];
        if (hooks) {
            hooks.forEach((hook) => hook(...args));
        }
        if (originLifecycle !== undefined) {
            originLifecycle.call(this, ...args);
        }
    };
}

const pageHookWarn = 'Page specific lifecycle injection APIs can only be used during execution of setup() in definePage() or defineComponent().';
const onAppShow = createAppHook(AppLifecycle.ON_SHOW);
const onAppHide = createAppHook(AppLifecycle.ON_HIDE);
const onAppError = createAppHook(AppLifecycle.ON_ERROR);
const onPageNotFound = createAppHook(AppLifecycle.ON_PAGE_NOT_FOUND);
const onUnhandledRejection = createAppHook(AppLifecycle.ON_UNHANDLED_REJECTION);
const onThemeChange = createAppHook(AppLifecycle.ON_THEME_CHANGE);
const onShow = createPageHook(PageLifecycle.ON_SHOW);
const onHide = createPageHook(PageLifecycle.ON_HIDE);
const onUnload = createPageHook(PageLifecycle.ON_UNLOAD);
const onRouteDone = createPageHook(PageLifecycle.ON_ROUTE_DONE);
const onPullDownRefresh = createPageHook(PageLifecycle.ON_PULL_DOWN_REFRESH);
const onReachBottom = createPageHook(PageLifecycle.ON_REACH_BOTTOM);
const onResize = createPageHook(PageLifecycle.ON_RESIZE);
const onTabItemTap = createPageHook(PageLifecycle.ON_TAB_ITEM_TAP);
const onPageScroll = (hook) => {
    const currentInstance = getCurrentInstance();
    /* istanbul ignore else -- @preserve  */
    if (currentInstance) {
        /* istanbul ignore else -- @preserve   */
        if (currentInstance.__listenPageScroll__) {
            injectHook(currentInstance, PageLifecycle.ON_PAGE_SCROLL, hook);
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('onPageScroll() hook only works when `listenPageScroll` is configured to true.');
        }
    }
    else if ((process.env.NODE_ENV !== 'production')) {
        console.warn(pageHookWarn);
    }
};
const onShareAppMessage = (hook) => {
    const currentInstance = getCurrentInstance();
    /* istanbul ignore else -- @preserve  */
    if (currentInstance) {
        /* istanbul ignore else -- @preserve  */
        if (currentInstance[PageLifecycle.ON_SHARE_APP_MESSAGE] &&
            currentInstance.__isInjectedShareToOthersHook__) {
            const hiddenField = toHiddenField(PageLifecycle.ON_SHARE_APP_MESSAGE);
            /* istanbul ignore else -- @preserve  */
            if (currentInstance[hiddenField] === undefined) {
                currentInstance[hiddenField] = hook;
            }
            else if ((process.env.NODE_ENV !== 'production')) {
                console.warn('onShareAppMessage() hook can only be called once.');
            }
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('onShareAppMessage() hook only works when `onShareAppMessage` option is not exist and `canShareToOthers` is configured to true.');
        }
    }
    else if ((process.env.NODE_ENV !== 'production')) {
        console.warn(pageHookWarn);
    }
};
const onShareTimeline = (hook) => {
    const currentInstance = getCurrentInstance();
    /* istanbul ignore else -- @preserve  */
    if (currentInstance) {
        /* istanbul ignore else -- @preserve  */
        if (currentInstance[PageLifecycle.ON_SHARE_TIMELINE] &&
            currentInstance.__isInjectedShareToTimelineHook__) {
            const hiddenField = toHiddenField(PageLifecycle.ON_SHARE_TIMELINE);
            /* istanbul ignore else -- @preserve  */
            if (currentInstance[hiddenField] === undefined) {
                currentInstance[hiddenField] = hook;
            }
            else if ((process.env.NODE_ENV !== 'production')) {
                console.warn('onShareTimeline() hook can only be called once.');
            }
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('onShareTimeline() hook only works when `onShareTimeline` option is not exist and `canShareToTimeline` is configured to true.');
        }
    }
    else if ((process.env.NODE_ENV !== 'production')) {
        console.warn(pageHookWarn);
    }
};
const onAddToFavorites = (hook) => {
    const currentInstance = getCurrentInstance();
    /* istanbul ignore else -- @preserve  */
    if (currentInstance) {
        /* istanbul ignore else -- @preserve  */
        if (currentInstance.__isInjectedFavoritesHook__) {
            const hiddenField = toHiddenField(PageLifecycle.ON_ADD_TO_FAVORITES);
            /* istanbul ignore else -- @preserve  */
            if (currentInstance[hiddenField] === undefined) {
                currentInstance[hiddenField] = hook;
            }
            else if ((process.env.NODE_ENV !== 'production')) {
                console.warn('onAddToFavorites() hook can only be called once.');
            }
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('onAddToFavorites() hook only works when `onAddToFavorites` option is not exist.');
        }
    }
    else if ((process.env.NODE_ENV !== 'production')) {
        console.warn(pageHookWarn);
    }
};
const onSaveExitState = (hook) => {
    const currentInstance = getCurrentInstance();
    /* istanbul ignore else -- @preserve  */
    if (currentInstance) {
        /* istanbul ignore else -- @preserve  */
        if (currentInstance.__isInjectedExitStateHook__) {
            const hiddenField = toHiddenField(PageLifecycle.ON_SAVE_EXIT_STATE);
            /* istanbul ignore else -- @preserve  */
            if (currentInstance[hiddenField] === undefined) {
                currentInstance[hiddenField] = hook;
            }
            else if ((process.env.NODE_ENV !== 'production')) {
                console.warn('onSaveExitState() hook can only be called once.');
            }
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('onSaveExitState() hook only works when `onSaveExitState` option is not exist.');
        }
    }
    else if ((process.env.NODE_ENV !== 'production')) {
        console.warn(pageHookWarn);
    }
};
const onReady = (hook) => {
    const currentInstance = getCurrentInstance();
    /* istanbul ignore else -- @preserve  */
    if (currentInstance) {
        injectHook(currentInstance, PageLifecycle.ON_READY, hook);
    }
    else if ((process.env.NODE_ENV !== 'production')) {
        console.warn('onReady() hook can only be called during execution of setup() in definePage() or defineComponent().');
    }
};
const onLoad = createComponentHook(PageLifecycle.ON_LOAD);
const onMove = createComponentHook(ComponentLifecycle.MOVED);
const onDetach = createComponentHook(ComponentLifecycle.DETACHED);
const onError = createComponentHook(ComponentLifecycle.ERROR);
function createAppHook(lifecycle) {
    return (hook) => {
        /* istanbul ignore else -- @preserve  */
        if (currentApp) {
            injectHook(currentApp, lifecycle, hook);
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('App specific lifecycle injection APIs can only be used during execution of setup() in createApp().');
        }
    };
}
function createPageHook(lifecycle) {
    return (hook) => {
        const currentInstance = getCurrentInstance();
        /* istanbul ignore else -- @preserve  */
        if (currentInstance) {
            injectHook(currentInstance, lifecycle, hook);
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn(pageHookWarn);
        }
    };
}
function createComponentHook(lifecycle) {
    return (hook) => {
        /* istanbul ignore else -- @preserve  */
        if (currentComponent) {
            injectHook(currentComponent, lifecycle, hook);
        }
        else if ((process.env.NODE_ENV !== 'production')) {
            console.warn('Component specific lifecycle injection APIs can only be used during execution of setup() in defineComponent().');
        }
    };
}
function injectHook(currentInstance, lifecycle, hook) {
    const hiddenField = toHiddenField(lifecycle);
    if (currentInstance[hiddenField] === undefined) {
        currentInstance[hiddenField] = [];
    }
    currentInstance[hiddenField].push(hook);
}

export { createApp, defineComponent, definePage, inject, nextTick, onAddToFavorites, onAppError, onAppHide, onAppShow, onDetach, onError, onHide, onLoad, onMove, onPageNotFound, onPageScroll, onPullDownRefresh, onReachBottom, onReady, onResize, onRouteDone, onSaveExitState, onShareAppMessage, onShareTimeline, onShow, onTabItemTap, onThemeChange, onUnhandledRejection, onUnload, provide, watch, watchEffect, watchPostEffect, watchSyncEffect };
